zerver/models.py

```diff
@@ -698,6 +698,41 @@ class RealmFilter(models.Model):
     class Meta:
         unique_together = ("realm", "pattern")
 
+    def clean(self) -> None:
+        # extract substitutions
+        try:
+            pattern = re.compile(self.pattern)
+        except Exception:
+            # regex validation errors would be caught by full_clean()
+            # so we can safely just return here.
+            return
+        group_list = list(pattern.groupindex.keys())
+        group_match_regex = re.compile(r'%\((?P<group_name>[^)]+)\)s')
+
+        # Report patterns missing in filter pattern.
+        for m in group_match_regex.finditer(self.url_format_string):
+            name = m.group('group_name')
+            if name not in group_list:
+                raise ValidationError(_("Group '{}' in URL format string is "
+                                        "not present in filter pattern.".format(name)))
+            group_list.remove(name)
+
+        # Report patterns missing in URL format string.
+        if len(group_list) == 1:
+            raise ValidationError(_("Group '{}' in filter pattern is not present "
+                                    "in URL format string.".format(group_list[0])))
+        if len(group_list) > 1:
+            group_list.sort()
+            raise ValidationError(_("Groups '{}' in filter pattern is not present "
+                                    "in URL format string.".format(', '.join(group_list))))
```

Comment:
I think it's an unnecessary cost to translators to have two variants here; it's extra work for translators.  I don't think it's critical to list all groups; so I'd prefer to refactor these blocks to deduplicate the strings for translator convenience (and just raise about the first item in `group_list`.

Reply:
Fixed.